src/hssm/hssm.py

```diff
@@ -684,25 +707,147 @@ def vi(
+            A pandas DataFrame with values for the predictors that are used to obtain
+            out-of-sample predictions. If omitted, the original dataset is used.
+        inplace : optional
+            If `True` will modify idata in-place and append a `log_likelihood` group to
+            `idata`. Otherwise, it will return a copy of idata with the predictions
+            added, by default True.
+        keep_likelihood_params : optional
+            If `True`, the trial wise likelihood parameters that are computed
+            on route to getting the log likelihood are kept in the `idata` object.
+            Defaults to False. See also the method `add_likelihood_parameters_to_idata`.
+
+        Returns
+        -------
+        az.InferenceData | None
+            InferenceData or None
+        """
+        if self._inference_obj is None and idata is None:
+            raise ValueError(
+                "Neither has the model been sampled yet nor"
+                + " an idata object has been provided."
+            )
+
+        if idata is None:
+            if self._inference_obj is None:
+                raise ValueError(
+                    "The model has not been sampled yet. "
+                    + "Please provide an idata object."
+                )
+            else:
+                idata = self._inference_obj
+
+        # Actual likelihood computation
+        idata = _compute_log_likelihood(self.model, idata, data, inplace)
+
+        # clean up posterior:
+        if not keep_likelihood_params:
+            self._clean_posterior_group(idata=idata)
+
+        if inplace:
+            return None
+        else:
+            return idata
+
+    def add_likelihood_parameters_to_idata(
+        self,
+        idata: az.InferenceData | None = None,
+        inplace: bool = False,
+    ) -> az.InferenceData | None:
+        """Add likelihood parameters to the InferenceData object.
+
+        Parameters
+        ----------
+        idata : az.InferenceData
+            The InferenceData object returned by HSSM.sample().
+        inplace : bool
+            If True, the likelihood parameters are added to idata in-place. Otherwise,
+            a copy of idata with the likelihood parameters added is returned.
+            Defaults to False.
+
+        Returns
+        -------
+        az.InferenceData | None
+            InferenceData or None
+        """
+        if idata is None and self._inference_obj is not None:
+            idata = self.model._compute_likelihood_params(  # pylint: disable=protected-access
+                deepcopy(self._inference_obj) if not inplace else self._inference_obj
+            )
+        elif idata is None and self._inference_obj is None:
+            raise ValueError("No idata provided and model not yet sampled!")
+        elif idata is not None:
+            idata = self.model._compute_likelihood_params(  # pylint: disable=protected-access
+                deepcopy(idata) if not inplace else idata
```

Comment:
To the best of my understanding the situation is:

Shallow copy will create a new combound object (az.InferenceData), but only references to components. If you make a change to a referenced component this will change the original. 

That can sometimes happen here and it could create weird side-effects when people expect decoupled objects as a result of not doing things in place. 

If I a wrong about this, let me know, but otherwise prefer to keep deepcopy here (agree annoying because expensive, however since this is done only once, I actually think it will never be a very serious issue, if anything RAM issues possibly appear as a result of this, but in that case the `inplace=True` option is around.

Reply:
Nvm. Just realized that ArviZ does not provide a shallow copy method (`InferenceData.copy()` actually just a call to `deepcopy` anyway, which I think is a very bad idea because people would naturally think it's like a shallow copy like how it is implemented in `dict` and `numpy`).

To this problem specifically, a shallow copy should not cause problems imo. Unless people perform an array-level operation, like indexing into an underlying array in the InferenceData and change values, which I haven't seen anyone doing with an InferenceData object, things should be fine